File: hub/src/light/effect/translation_effects/constant_targets_effect.py
from light.effect.partial_effect import PartialEffect
from light.led_selector import LEDSelector
from light.effect.resolved_effect import ResolvedEffect

class ConstantTargetsEffect(PartialEffect):
    def __init__(self, targets, dt=0):
        super().__init__(dt)
        self.targets = targets

    def getEffect(self, t):
        pos = []
        for target in self.targets:
            pos.extend(LEDSelector().getAllTargetPositions(target))

        return ResolvedEffect.createTranslation(list(set(pos)))

    def isTemporal(self):
        # Constant targets do not control timing
        return False

    def isComplete(self, t):
        return True

File: hub/drumlite-js/src/light/effect/positional/ConstantTargetsEffect.ts
import PartialEffect, { EffectParameters, EffectParameter } from "../PartialEffect"
import { EffectTarget } from "../EffectTarget"
import ResolvedEffect from "../../../effect/ResolvedEffect";
import LEDSelector from "../../LEDSelector";

export class ConstantTargetsEffectParams extends EffectParameters {
    effectName = "Constant Targets";
    typeName = "Positional";

    constructor(targets: EffectTarget[] = []) {
        super();
        this.params.targets = new EffectParameter<EffectTarget[]>("Targets", targets, {type: "target", isArray: true})
    }
}

export default class ConstantTargetsEffect extends PartialEffect<ConstantTargetsEffectParams> {
    constructor(params: ConstantTargetsEffectParams, dt: number = 0) {
        super(params, dt);
    }

    public getEffect(t: number) {
        const pos = [];
        for (let target of (this.params.params.targets.val)) {
            pos.push(...new LEDSelector().getAllTargetPositions(target));
        }

        return [[ResolvedEffect.createTranslation(Array.from(new Set<number>(pos)))]];
    }

    public isTemporal() {
        // Constant targets do not control timing
        return false;
    }

    public isComplete(t: number) {
        return true;
    }
}